src/devices/bus/spectrum/nemoide.cpp

```diff
@@ -0,0 +1,91 @@
+// license:BSD-3-Clause
+// copyright-holders:Andrei I. Holub
+
+#include "emu.h"
+#include "nemoide.h"
+
+#include "bus/ata/ataintf.h"
+#include "bus/ata/atapicdr.h"
+#include "bus/ata/hdd.h"
+
+namespace bus::spectrum::zxbus {
+
+namespace {
+
+class nemoide_device : public device_t, public device_zxbus_card_interface
+{
+public:
+	nemoide_device(const machine_config &mconfig, const char *tag, device_t *owner, u32 clock)
+		: device_t(mconfig, ZXBUS_NEMOIDE, tag, owner, clock)
+		, device_zxbus_card_interface(mconfig, *this)
+		, m_ata(*this, "ata")
+	{ }
+
+protected:
+	virtual void device_add_mconfig(machine_config &config) override;
+	virtual void device_start() override;
+	virtual void device_reset() override;
+
+private:
+	void map_io(address_map &map);
+
+	u8 ata_r(offs_t offset);
+	void ata_w(offs_t offset, u8 data);
+
+	required_device<ata_interface_device> m_ata;
+
+	u8 m_ata_data_latch;
+};
+
+u8 nemoide_device::ata_r(offs_t offset)
+{
+    const u16 data = m_ata->cs0_r(offset & 7);
+	if (!machine().side_effects_disabled())
+		m_ata_data_latch = data >> 8;
+
+	return data & 0xff;
+}
+
+void nemoide_device::ata_w(offs_t offset, u8 data)
+{
+	const u16 ata_data = (m_ata_data_latch << 8) | data;
+    m_ata->cs0_w(offset & 7, ata_data);
+}
+
+void nemoide_device::map_io(address_map &map)
+{
+	map(0x00011, 0x00011).mirror(0xff00).lrw8(NAME([this]() { return m_ata_data_latch; })
+		, NAME([this](offs_t offset, u8 data) { m_ata_data_latch = data; }));
+	map(0x00010, 0x00010).select(0xffe0).lrw8(NAME([this](offs_t offset) { return ata_r(offset >> 5); })
+		, NAME([this](offs_t offset, u8 data) { ata_w(offset >> 5, data); }));
+	map(0x000c8, 0x000c8).mirror(0xff00).lrw8(NAME([this]() { return m_ata->cs1_r(6); })
+		, NAME([this](offs_t offset, u8 data) { m_ata->cs1_w(6, data); }));
```

Comment:
Why have the shifts here, rather than doing them in `ata_r` and `ata_w`?  Those methods aren’t used anywhere else anyway, and putting the shifts here clutters the address map.